docs/book/v3/async-tasks.md

```diff
@@ -0,0 +1,888 @@
+
+First, let's define an event type:
+
+```php
+namespace Example;
+
+class SomeDeferrableEvent
+{
+    /** @var string */
+    private $message;
+
+    public function __construct(string $message)
+    {
+        $this->message = $message;
+    }
+
+    public function __toString(): string
+    {
+        return $this->message;
+    }
+}
+```
+
+Next, we will create a listener for this event type:
+
+```php
+namespace Example;
+
+use Psr\Log\LoggerInterface;
+
+class SomeDeferrableEventListener
+{
+    /** @var LoggerInterface */
+    private $logger;
+
+    public function __construct(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
+    }
+
+    public function __invoke(SomeDeferrableEvent $event): void
+    {
+        $this->log->info(sprintf('Message: %s', (string) $event));
+    }
+}
+```
+
+The listener will require a factory:
+
+```php
+namespace Example;
+
+use Psr\Container\ContainerInterface;
+use Psr\Log\LoggerInterface;
+
+class SomeDeferrableEventListenerFactory
+{
+    public function __invoke(ContainerInterface $container): SomeDeferrableEventListener
+    {
+        return new SomeDeferrableEventListener(
+            $container->get(LoggerInterface::class)
+        );
+    }
+}
+```
+
+At this point, we turn to configuration.
+We will add dependency configuration for our listener (omitting the configuration for the logger service; we will assume you have done so already, or can figure out how to do so).
+We will also add configuration to bind the `TaskInvokerListener` to the `TaskEvent`.
+
+```php
+// in config/autoload/swoole.global.php or similar:
+
+use Example\SomeDeferrableEventListener;
+use Example\SomeDeferrableEventListenerFactory;
+use Mezzio\Swoole\Event\TaskEvent;
+use Mezzio\Swoole\Event\TaskInvokerListener;
+
+return [
+    'dependencies' => [
+        'factories' => [
+            SomeDeferrableEventListener::class => SomeDeferrableEventListenerFactory::class,
+        ],
+    ],
+    'mezzio-swoole' => [
+        'swoole-http-server' => [
+            'listeners' => [
+                TaskEvent::class => [
+                    TaskInvokerListener::class,
+                ],
+            ],
+        ],
+    ],
+];
+```
+
+Next, we will create a `ServiceBasedTask` instance.
+The constructor for `Mezzio\Swoole\Task\Task` expects the name of a service that can be pulled from the application DI container as the first argument, and then zero or more additional arguments representing the arguments to pass to the listener.
+The service is expected to be invokable (i.e., it MUST define the method `__invoke()`).
+
+```php
+use Example\SomeDeferrableEvent;
+use Example\SomeDeferrableEventListener;
+use Mezzio\Swoole\Task\ServiceBasedTask;
+
+$task = new ServiceBasedTask(
+    SomeDeferrableEventListener::class,
+    new SomeDeferrableEvent('hello world')
+);
+```
+
```

Comment:
Confused as to why a SomeDeferrableEvent with a *listener* is created, and not a concrete SomeDeferrableTask invoked by the TaskInvokerListener?

Reply:
This is a lead up to the next section, where the code that creates it will be _dispatching an event_, not creating a task.

Basically, this whole write-up is leading up to the idea that instead of manually and explicitly queuing tasks with the server, you will dispatch events, and have listeners opt-in to whether they execute asynchronously via task workers by adding the `DeferredListenerDelegator` or `DeferredServiceListenerDelegator`.